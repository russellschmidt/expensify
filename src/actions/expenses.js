import uuid from 'uuid'
import database from '../firebase/firebase'

// v1
// component calls action generator
// action generator returns object
// component dispatches object
// redux store changes

// v2 w firebase
// component calls action generator
// action generator returns function
// component dispatches function - redux doesnt allow so we need a module
// function runs - push to firebase and dispatch action to update store

export const addExpense = (expense) => ({
  type: 'ADD_EXPENSE',
  expense
})

export const startAddExpense = (expenseData = {}) => {
  return (dispatch) => {
    const {
      description = '', 
      note = '', 
      amount = 0, 
      createdAt = 0
    } = expenseData

    const expense = {description, note, amount, createdAt}

    return database.ref('expenses').push(expense).then((ref) => {
      dispatch(addExpense({
        id: ref.key,
        ...expense
      }))
    })
  }
}

// Remove Expense
export const removeExpense = ({id} = {}) => ({
  type: 'REMOVE_EXPENSE',
  id
})

export const startRemoveExpense = ({id} = {}) => {
  return (dispatch) => {
    return database.ref(`expenses/${id}`).remove().then(() => {
      dispatch(removeExpense({id}))
    })
  }
}

// Edit Expense
export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
})

// set_expenses 
export const setExpenses = (expenses) => ({
  type: 'SET_EXPENSES',
  expenses
})

// export const startSetExpenses

export const startSetExpenses = () => {
  return (dispatch) => {
    return database.ref('expenses').once('value').then((snapshot) => {
      const expenses = []
      snapshot.forEach((childSnapshot) => {
        expenses.push({
          id: childSnapshot.key,
          ...childSnapshot.val()
        })
      })
      dispatch(setExpenses(expenses))
    })
  }
}