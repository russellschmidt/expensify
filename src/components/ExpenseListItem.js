import React from 'react'
import {Link} from 'react-router-dom'

export const ExpenseListItem = (props) => (
  <tr>
    <td><Link to={`/edit/${props.id}`}>{props.description}</Link></td>
    <td>${(props.amount / 100).toFixed(2)}</td>
    <td>{props.createdAt}</td>
  </tr>
)

export default ExpenseListItem