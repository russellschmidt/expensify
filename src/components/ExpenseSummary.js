import React from 'react'
import { connect } from 'react-redux'
import numeral from 'numeral'
import {Link} from 'react-router-dom'
import selectExpenses from '../selectors/expenses'
import getExpenseTotal from '../selectors/expenses-total'

export const ExpenseSummaryPage = (props) => {
  const formattedExpense = numeral(getExpenseTotal(props.expenses)).format('$0, 0.00')
  const expenseCount = props.expenses.length
  return (
    <div className="page-header">
      <div className="content-container">
        <h1 className="page-header__title">Viewing <span>{expenseCount}</span> expense{ expenseCount === 1 ? '' : 's'} totaling <span>{formattedExpense}</span></h1>
        <div className="page-header__actions">
          <Link className="page-header__button" to="/create">Add Expense</Link>
        </div>
      </div> 
    </div>
    )
  }

const mapStateToProps = (state, props) => ({
  expenses: selectExpenses(state.expenses, state.filters)
})

export default connect(mapStateToProps)(ExpenseSummaryPage)
