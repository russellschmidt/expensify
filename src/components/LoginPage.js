import React from 'react'

const LoginPage = () => (
  <div>
    <button>Log in</button>
  </div>
)

export { LoginPage }