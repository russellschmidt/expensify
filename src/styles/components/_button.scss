.box-layout__button {
  background: $deep-blue;
  border: none;
  border-radius: 4px;
  box-shadow: 2px 2px 3px $shadow-grey;
  color: white;
  font-size: $font-size-large;
  font-weight: 300;
  height: 5rem;
  width: 20rem;
}