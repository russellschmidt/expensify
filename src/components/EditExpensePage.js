import React from 'react'
import {connect} from 'react-redux'
import ExpenseForm from './ExpenseForm'
import {editExpense, removeExpense} from '../actions/expenses'

export class EditExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.editExpense(this.props.expense.id, expense)
    this.props.history.push('/')
  }
  onRemove = (expense) => {
    this.props.removeExpense(this.props.expense)
    this.props.history.push('/')
  }
  render() {
    return(
      <div>
        <h1>Edit Expense</h1>
        <ExpenseForm 
          expense={this.props.expense} 
          onSubmit={this.onSubmit}
        />
        <button onClick={this.onRemove}>
          Remove
        </button>
      </div>
    )
  }
}

const mapStateToProps = (state, props) => ({
    expense: state.expenses.find((expense) => expense.id === props.match.params.id)
  }
)

const mapDispatchToProps = (dispatch, props) => {
  return {
    editExpense: (id, expense) => dispatch(editExpense(id, expense)),
    removeExpense: (expense) => dispatch(removeExpense(expense))
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage)

// const EditExpensePage = (props) => {
//   return (
//     <div>
//       <ExpenseForm 
//         expense={props.expense}
//         onSubmit={(expense) => {
//           props.dispatch(editExpense(props.expense.id, expense))
//           props.history.push('/')
//         }}
//       />

//       <button onClick={() => {
//         props.dispatch(removeExpense(props.expense))
//         props.history.push('/')
//       }}>Remove</button>
//     </div>
//   )
// }

// const MapStateToProps = (state, props) => {
//   return {
//     expense: state.expenses.find((expense) => expense.id === props.match.params.id)
//   } 
// }


// export default connect(MapStateToProps)(EditExpensePage)