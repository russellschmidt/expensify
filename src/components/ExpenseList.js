import React from 'react'
import { connect } from 'react-redux'
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses'

const ExpenseList = (props) => (
  <div>
    <h1>Expense List</h1>
    <table>
      <thead>
        <tr>
          <th>Description</th>
          <th>Amount</th>
          <th>Created At</th>
        </tr>
      </thead>
      <tbody>
        {props.expenses.map(expense => <ExpenseListItem {...expense} key={expense.id}/>)}
      </tbody>
    </table>
  </div>
) 

const mapStateToProps = (state) => {
  return {
    expenses: selectExpenses(state.expenses, state.filters)
  }
}

// ConnectedExpenseList
export default connect(mapStateToProps)(ExpenseList)