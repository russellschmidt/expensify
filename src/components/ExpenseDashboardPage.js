import React from 'react'
import ExpenseListFilters from './ExpenseListFilters'
import ExpenseSummary from './ExpenseSummary'
import ExpenseList from './ExpenseList'

const ExpenseDashboardPage = () => (
  <div>
    <ExpenseSummary />
    <ExpenseListFilters />
    <ExpenseList />
  </div>
)

export default ExpenseDashboardPage