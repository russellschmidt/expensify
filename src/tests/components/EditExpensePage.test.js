import React from 'react'
import { shallow } from 'enzyme'
import { EditExpensePage } from '../../components/EditExpensePage'
import expenses from '../fixtures/expenses'

let editExpense, history, wrapper

beforeEach(() => {
  editExpense = jest.fn()
  history = {push: jest.fn()}
  wrapper = shallow(<EditExpensePage editExpense={onSubmit} history={history}/>)
})

test('should render AddExpensePage correctly', () => {
  expect(wrapper).toMatchSnapshot()
})

test('should handle onSubmit', () => {
  wrapper.find('ExpenseForm').prop('onSubmit')(expenses[1])
  expect(history.push).toHaveBeenLastCalledWith('/')
  expect(addExpense).toHaveBeenLastCalledWith(expenses[1]) 
})