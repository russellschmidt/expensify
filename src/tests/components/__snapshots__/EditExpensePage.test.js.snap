// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditExpensePage correctly 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1>
        Edit Expense
      </h1>
      <ExpenseForm
        expense={
          Object {
            "amount": 109500,
            "createdAt": -345600000,
            "description": "rent",
            "id": "2b",
            "note": "hey",
          }
        }
        onSubmit={[Function]}
      />
      <button
        className="remove-expense__button"
        onClick={[Function]}
      >
        Remove Expense
      </button>
    </div>
  </div>
</div>
`;
